File: ts-api/examples/example_infill.ts
/**
 * NovelAI API Integration Tests
 * 実際にAPIリクエストを行うテスト
 * 
 * 使用方法: pnpm tsx test.ts
 */

import { NovelAIClient } from '../src/client';
import dotenv from 'dotenv';
import fs from 'fs';
import path from 'path';

dotenv.config();

const client = new NovelAIClient();
const INPUT_IMAGE = './reference/input.jpeg';
const OUTPUT_DIR = './output/test/';

// 出力ディレクトリ作成
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

/**
 * Test 1: Img2Img (Image to Image) - 元画像を参照して新しい画像を生成
 */
async function testImg2Img() {
  console.log('\n=== Test: Img2Img ===');
  try {
    const result = await client.generate({
      prompt: '1girl, beautiful, masterpiece',
      action: 'img2img',
      source_image: INPUT_IMAGE,
      img2img_strength: 0.6,
      img2img_noise: 0.1,
      width: 832,
      height: 1216,
      save_dir: OUTPUT_DIR,
    });
    console.log('✅ Img2Img success!');
    console.log(`   Saved to: ${result.saved_path}`);
    console.log(`   Anlas consumed: ${result.anlas_consumed}`);
    return true;
  } catch (error) {
    console.error('❌ Img2Img failed:', error);
    return false;
  }
}

/**
 * Test 2: Infill (マスクのみ) - マスク領域のみを再生成
 * ※ マスク画像が必要（白=再生成エリア, 黒=保持エリア）
 */
async function testInfillOnly() {
  console.log('\n=== Test: Infill (Mask Only) ===');

  // 簡易マスク画像を作成（中央部分を白にする）
  const maskPath = path.join(OUTPUT_DIR, 'temp_mask.png');

  // sharpがなければスキップ
  try {
    const sharp = (await import('sharp')).default;

    // 832x1216 の画像、中央600x800を白、それ以外を黒
    const mask = await sharp({
      create: {
        width: 832,
        height: 1216,
        channels: 4,
        background: { r: 0, g: 0, b: 0, alpha: 255 }
      }
    })
      .composite([{
        input: Buffer.from(
          `<svg width="832" height="1216">
          <rect x="116" y="208" width="600" height="800" fill="white"/>
        </svg>`
        ),
        top: 0,
        left: 0
      }])
      .png()
      .toBuffer();

    fs.writeFileSync(maskPath, mask);
    console.log(`   Created temp mask: ${maskPath}`);

    const result = await client.generate({
      prompt: '1girl, smiling, happy',
      action: 'infill',
      source_image: INPUT_IMAGE,
      mask: maskPath,
      mask_strength: 0.7,  // マスク反映度
      width: 832,
      height: 1216,
      save_dir: OUTPUT_DIR,
    });
    console.log('✅ Infill (Mask Only) success!');
    console.log(`   Saved to: ${result.saved_path}`);
    console.log(`   Anlas consumed: ${result.anlas_consumed}`);

    // 一時ファイル削除
    fs.unlinkSync(maskPath);
    return true;
  } catch (error) {
    console.error('❌ Infill (Mask Only) failed:', error);
    return false;
  }
}

/**
 * Test 3: Infill + Img2Img (ハイブリッドモード) - マスクと元画像参照を同時使用
 */
async function testInfillWithImg2Img() {
  console.log('\n=== Test: Infill + Img2Img (Hybrid Mode) ===');

  const maskPath = path.join(OUTPUT_DIR, 'temp_mask_hybrid.png');

  try {
    const sharp = (await import('sharp')).default;

    // マスク画像作成
    const mask = await sharp({
      create: {
        width: 832,
        height: 1216,
        channels: 4,
        background: { r: 0, g: 0, b: 0, alpha: 255 }
      }
    })
      .composite([{
        input: Buffer.from(
          `<svg width="832" height="1216">
          <rect x="116" y="208" width="600" height="800" fill="white"/>
        </svg>`
        ),
        top: 0,
        left: 0
      }])
      .png()
      .toBuffer();

    fs.writeFileSync(maskPath, mask);
    console.log(`   Created temp mask: ${maskPath}`);

    const result = await client.generate({
      prompt: '1girl, beautiful dress, elegant',
      action: 'infill',
      source_image: INPUT_IMAGE,
      mask: maskPath,
      mask_strength: 0.68,              // マスク反映度
      hybrid_img2img_strength: 0.45,    // 元画像維持率
      hybrid_img2img_noise: 0,          // 元画像ノイズ
      width: 832,
      height: 1216,
      save_dir: OUTPUT_DIR,
    });
    console.log('✅ Infill + Img2Img (Hybrid) success!');
    console.log(`   Saved to: ${result.saved_path}`);
    console.log(`   Anlas consumed: ${result.anlas_consumed}`);

    // 一時ファイル削除
    fs.unlinkSync(maskPath);
    return true;
  } catch (error) {
    console.error('❌ Infill + Img2Img (Hybrid) failed:', error);
    return false;
  }
}

// メイン実行
async function main() {
  console.log('========================================');
  console.log('NovelAI API Integration Tests');
  console.log(`Input Image: ${INPUT_IMAGE}`);
  console.log(`Output Dir: ${OUTPUT_DIR}`);
  console.log('========================================');

  // 入力画像存在チェック
  if (!fs.existsSync(INPUT_IMAGE)) {
    console.error(`❌ Input image not found: ${INPUT_IMAGE}`);
    process.exit(1);
  }

  const results: { name: string; success: boolean }[] = [];

  // Test 1: Img2Img
  results.push({ name: 'Img2Img', success: await testImg2Img() });

  // Test 2: Infill (Mask Only)
  results.push({ name: 'Infill (Mask Only)', success: await testInfillOnly() });

  // Test 3: Infill + Img2Img (Hybrid)
  results.push({ name: 'Infill + Img2Img', success: await testInfillWithImg2Img() });

  // 結果サマリー
  console.log('\n========================================');
  console.log('Test Results Summary');
  console.log('========================================');
  for (const r of results) {
    console.log(`${r.success ? '✅' : '❌'} ${r.name}`);
  }

  const failed = results.filter(r => !r.success).length;
  if (failed > 0) {
    console.log(`\n${failed} test(s) failed.`);
    process.exit(1);
  } else {
    console.log('\nAll tests passed!');
  }
}

main();

File: rust-api/examples/example_infill.rs
//! NovelAI Infill (Inpaint) Example
//!
//! マスクを使った部分再生成のサンプル
//!
//! Run with: cargo run --example example_infill

use std::io::Cursor;
use std::path::Path;

use anyhow::Result;
use image::{ImageEncoder, RgbaImage};
use novelai_api::client::NovelAIClient;
use novelai_api::schemas::*;

/// Create a PNG mask image in memory.
/// White rectangle = area to regenerate, black = area to keep.
fn create_mask(
    width: u32,
    height: u32,
    rect_x: u32,
    rect_y: u32,
    rect_w: u32,
    rect_h: u32,
) -> Result<Vec<u8>> {
    let img = RgbaImage::from_fn(width, height, |x, y| {
        if x >= rect_x && x < rect_x + rect_w && y >= rect_y && y < rect_y + rect_h {
            image::Rgba([255, 255, 255, 255]) // White = regenerate
        } else {
            image::Rgba([0, 0, 0, 255]) // Black = keep
        }
    });

    let mut buf = Cursor::new(Vec::new());
    let encoder = image::codecs::png::PngEncoder::new(&mut buf);
    encoder.write_image(
        img.as_raw(),
        width,
        height,
        image::ExtendedColorType::Rgba8,
    )?;

    Ok(buf.into_inner())
}

/// Test 1: Basic img2img (baseline)
async fn test_img2img(client: &NovelAIClient, input_image: &str, output_dir: &str) -> Result<()> {
    println!("\n=== Test: Img2Img ===");

    let params = GenerateParams::builder("1girl, beautiful, masterpiece")
        .action(GenerateAction::Img2Img {
            source_image: ImageInput::FilePath(input_image.into()),
            strength: 0.6,
            noise: 0.1,
        })
        .width(832)
        .height(1216)
        .save_dir(output_dir)
        .build()?;

    let result = client.generate(&params).await?;

    println!(
        "  Img2Img success! Saved to: {}",
        result.saved_path.as_deref().unwrap_or("N/A")
    );
    println!(
        "  Anlas consumed: {}",
        result
            .anlas_consumed
            .map_or("N/A".into(), |v| v.to_string())
    );

    Ok(())
}

/// Test 2: Infill + Img2Img hybrid mode
async fn test_infill_with_img2img(
    client: &NovelAIClient,
    input_image: &str,
    output_dir: &str,
) -> Result<()> {
    println!("\n=== Test: Infill + Img2Img (Hybrid Mode) ===");

    let mask_bytes = create_mask(832, 1216, 116, 208, 600, 800)?;
    println!("  Created in-memory mask (832x1216, rect at 116,208 600x800)");

    let params = GenerateParams::builder("1girl, beautiful dress, elegant")
        .action(GenerateAction::Infill {
            source_image: ImageInput::FilePath(input_image.into()),
            mask: ImageInput::Bytes(mask_bytes),
            mask_strength: 0.68,
            color_correct: false,
            hybrid_strength: Some(0.45),
            hybrid_noise: Some(0.0),
        })
        .width(832)
        .height(1216)
        .save_dir(output_dir)
        .build()?;

    let result = client.generate(&params).await?;

    println!(
        "  Infill + Img2Img (Hybrid) success! Saved to: {}",
        result.saved_path.as_deref().unwrap_or("N/A")
    );
    println!(
        "  Anlas consumed: {}",
        result
            .anlas_consumed
            .map_or("N/A".into(), |v| v.to_string())
    );

    Ok(())
}

#[tokio::main]
async fn main() -> Result<()> {
    dotenvy::dotenv().ok();

    let input_image = "./reference/input.jpeg";
    let output_dir = "./output/test/";

    std::fs::create_dir_all(output_dir)?;

    println!("========================================");
    println!("NovelAI API Infill Tests");
    println!("Input Image: {}", input_image);
    println!("Output Dir: {}", output_dir);
    println!("========================================");

    if !Path::new(input_image).exists() {
        eprintln!("Input image not found: {}", input_image);
        std::process::exit(1);
    }

    let client = NovelAIClient::new(None, None)?;

    let mut results: Vec<(&str, bool)> = Vec::new();

    // Test 1: Img2Img
    let success = test_img2img(&client, input_image, output_dir).await.is_ok();
    results.push(("Img2Img", success));

    // Test 2: Infill + Img2Img (Hybrid)
    let success = test_infill_with_img2img(&client, input_image, output_dir)
        .await
        .is_ok();
    results.push(("Infill + Img2Img", success));

    // Summary
    println!("\n========================================");
    println!("Test Results Summary");
    println!("========================================");
    for (name, success) in &results {
        let icon = if *success { "OK" } else { "NG" };
        println!("{} {}", icon, name);
    }

    let failed = results.iter().filter(|(_, s)| !*s).count();
    if failed > 0 {
        println!("\n{} test(s) failed.", failed);
        std::process::exit(1);
    } else {
        println!("\nAll tests passed!");
    }

    Ok(())
}
